Source/WebCore/html/shadow/SpatialImageControls.cpp

```diff
@@ -0,0 +1,214 @@
+    imageElement.protectedDocument()->checkedEventLoop()->queueTask(TaskSource::InternalAsyncTask, [weakElement = WeakPtr { imageElement }] {
+        RefPtr element = weakElement.get();
+        RefPtr shadowRoot = element->shadowRoot();
+        ASSERT(shadowRoot);
+        Ref document = element->document();
+
+        if (hasSpatialImageControls(*element))
+            return;
+
+        Ref controlLayer = HTMLDivElement::create(document.get());
+        controlLayer->setIdAttribute(spatialImageControlsElementIdentifier());
+        controlLayer->setAttributeWithoutSynchronization(HTMLNames::contenteditableAttr, falseAtom());
+        controlLayer->setAttributeWithoutSynchronization(HTMLNames::styleAttr, "position: relative;"_s);
+        shadowRoot->appendChild(controlLayer);
+
+        static MainThreadNeverDestroyed<const String> shadowStyle(StringImpl::createWithoutCopying(spatialImageControlsUserAgentStyleSheet));
+        Ref style = HTMLStyleElement::create(HTMLNames::styleTag, document.get(), false);
+        style->setTextContent(String { shadowStyle });
+        controlLayer->appendChild(WTFMove(style));
+
+        Ref button = HTMLButtonElement::create(HTMLNames::buttonTag, document.get(), nullptr);
+        button->setIdAttribute(spatialImageControlsButtonIdentifier());
+        controlLayer->appendChild(button);
+
+        if (CheckedPtr renderImage = dynamicDowncast<RenderImage>(element->renderer()))
+            renderImage->setHasShadowControls(true);
+    });
+}
+
+bool handleEvent(HTMLElement& element, Event& event)
+{
+    if (!isAnyClick(event))
+        return false;
+
+    RefPtr frame = element.document().frame();
+    if (!frame)
+        return false;
+
+    RefPtr page = element.document().page();
+    if (!page)
+        return false;
+
+    RefPtr mouseEvent = dynamicDowncast<MouseEvent>(event);
+    if (!mouseEvent)
+        return false;
+
+    RefPtr target = dynamicDowncast<Element>(mouseEvent->target());
+    if (!target)
+        return false;
+
+    if (SpatialImageControls::isSpatialImageControlsButtonElement(*target)) {
+        RefPtr img = dynamicDowncast<HTMLImageElement>(target->shadowHost());
+        img->webkitRequestFullscreen();
+
+        event.setDefaultHandled();
+        return true;
+    }
+    return false;
+}
+
+void destroySpatialImageControls(HTMLElement& element)
+{
+    element.protectedDocument()->checkedEventLoop()->queueTask(TaskSource::InternalAsyncTask, [weakElement = WeakPtr { element }] {
+        RefPtr protectedElement = weakElement.get();
+        RefPtr shadowRoot = protectedElement->userAgentShadowRoot();
+        if (!shadowRoot)
+            return;
+
+        if (RefPtr node = shadowRoot->firstChild()) {
+            RefPtr htmlElement = dynamicDowncast<HTMLElement>(node.releaseNonNull());
+            if (!htmlElement)
+                return;
+            RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(htmlElement->getIdAttribute() == spatialImageControlsElementIdentifier());
```

Comment:
I don't think you can release-assert this. UA shadow tree could be hosting image overlay for example.

Reply:
I suggest we look for an element with the matching ID instead like this:
```cpp
if (RefPtr element = shadowRoot->getElementById(spatialImageControlsElementIdentifier()))
    element->remove();
```